sklearn/utils/_testing.py

```diff
@@ -747,99 +750,166 @@ def _convert_container(
-        return result
-    elif constructor_name == "pyarrow":
-        pa = pytest.importorskip("pyarrow", minversion=minversion)
-        array = np.asarray(container)
-        if columns_name is None:
-            columns_name = [f"col{i}" for i in range(array.shape[1])]
-        data = {name: array[:, i] for i, name in enumerate(columns_name)}
-        result = pa.Table.from_pydict(data)
-        if categorical_feature_names is not None:
-            for col_idx, col_name in enumerate(result.column_names):
-                if col_name in categorical_feature_names:
-                    result = result.set_column(
-                        col_idx, col_name, result.column(col_name).dictionary_encode()
-                    )
-        return result
-    elif constructor_name == "polars":
-        pl = pytest.importorskip("polars", minversion=minversion)
-        result = pl.DataFrame(container, schema=columns_name, orient="row")
-        if categorical_feature_names is not None:
-            for col_name in categorical_feature_names:
-                result = result.with_columns(pl.col(col_name).cast(pl.Categorical))
-        return result
-    elif constructor_name == "series":
-        pd = pytest.importorskip("pandas", minversion=minversion)
-        return pd.Series(container, dtype=dtype)
-    elif constructor_name == "polars_series":
-        pl = pytest.importorskip("polars", minversion=minversion)
-        return pl.Series(values=container)
-    elif constructor_name == "index":
-        pd = pytest.importorskip("pandas", minversion=minversion)
-        return pd.Index(container, dtype=dtype)
-    elif constructor_name == "slice":
+    if sp.sparse.issparse(container):
+        # Converting from a sparse container to a non-sparse container, so we densify
+        # the container in the first place
+        if sparse_container is None:
+            container = container.toarray()
+        if dtype is not None:
+            container = container.astype(dtype, copy=False)
+    else:
+        container = np.asarray(container, dtype=dtype)
+
+    # Check the dimension of the container
+    if constructor_type == "slice":
+        if container.shape != (2,):
+            raise ValueError(
+                "Only 1D containers with exactly 2 elements can be converted to slice; "
+                f"got shape {container.shape} instead"
+            )
+    elif constructor_type == "dataframe":
+        if container.ndim != 2:
+            raise ValueError(
+                "Only 2D containers can be converted to dataframe; got shape "
+                f"{container.shape} instead"
+            )
+    elif constructor_type in ("series", "index"):
+        if container.ndim != 1:
+            raise ValueError(
+                f"Only 1D containers can be converted to {constructor_type}; got shape "
+                f"{container.shape} instead"
+            )
+
+    # ---- Conversion ---------------------------------------------------------
+
```

Comment:
We usually don't use that kind of comment style. Since the function is getting pretty big, I think it could make sense to extract the individual converters for complex constructor types into dedicated smaller functions

Reply:
Thanks for the suggestion. I've created separate functions `_convert_container_to_{constructor_type}` for `"array"`, `"dataframe"`, `"series"`, and `"index"`. Those one-liners are left as is.